src/main/scala/mesosphere/util/ZookeeperDirectState.scala

```diff
@@ -0,0 +1,330 @@
+    ZKPaths.mkdirs(zkClient.getZookeeperClient.getZooKeeper, path)
+  }
+
+  private def exists(name: String): Boolean = {
+    client.checkExists().forPath(fullPath(name)) != null
+  }
+
+  implicit class LocalVariableConversion(v: Variable) {
+    def asLocal: LocalVariable = v match {
+      case lv: LocalVariable =>
+        lv
+      case _ =>
+        throw new IOException("An unexpected variable was encountered for a Zookeeper operation")
+    }
+  }
+
+}
+
+/**
+  * A trait that provides Apache Curator callbacks that convert to Scala promises,
+  * Scala futures and eventually Java futures. The Java futures implement a trait that
+  * embeds the original Scala future so that callers can access the underlying Scala future
+  * without creating new threads.
+  *
+  * Separated from ZookeeperDirectState to its own trait to keep the callback/promise/future features
+  * separate from the main State business.
+  */
+trait ZKBackgroundPromiseSupport { self: ZookeeperDirectState =>
+
+  /**
+    * A Callable implement that is not supposed to be called.
+    */
+  class Uncallable[A] extends Callable[A] {
+    override def call(): A = throw new IllegalArgumentException("The call() method is unsupported")
+  }
+
+  /**
+    * An inoperable Java Future. Typically used to contain a Scala Future using ScalaFutureSupport.
+    */
+  class UncallableJavaFuture[A] extends FutureTask[A](new Uncallable) {
+    override def get(l: Long, timeUnit: java.util.concurrent.TimeUnit): A = get()
+    override def get(): A = throw new IllegalArgumentException("get() is unsupported - call future() instead")
+  }
+
+  class ScalaFutureWrapper[A](val future: SFuture[A])
+    extends UncallableJavaFuture[A]
+    with ScalaFutureSupport[A]
+
+  implicit class ToScalaFutureWrapper[A](val future: SFuture[A]) {
+    def asScalaFutureWrapper = new ScalaFutureWrapper(future)
+  }
+
+  /**
+    * Base class for our combination Curator callbacks, Scala promises, Scala future containers
+    * and Java futures... all in one!
+    */
+  abstract class CallbackPromise[A] extends BackgroundCallback {
+    val prommy: Promise[A] = Promise[A]()
+    def future: SFuture[A] = prommy.future
+  }
+
+  /**
+    * A callback for zookeeper.getData() which returns it wrapped in a Future.
+    *
+    * Uses a Promise to generate a Scala Future. Also extends Java Future, but will throw an exception if
+    * called as a Java Future. This allows us to pass the Scala Future directly to Marathon/Chronos
+    * via the Java-Future-based State interface.
+    */
+  class GetDataCallbackPromise(name: String) extends CallbackPromise[Variable] {
+
+    override def processResult(client: CuratorFramework, event: CuratorEvent): Unit = {
+      val entry = StateProtos.Entry.parseFrom(event.getData)
+      val lv = new LocalVariable(name, entry.getValue.toByteArray, Some(entry.getUuid.toStringUtf8))
+      prommy.success(lv)
+    }
+
+    def success(v: Variable): Unit = { prommy.success(v) }
```

Comment:
Btw, this method shouldn't exist. Just create a completed future instead of the callback when it's not necessary.